ReactAndroid/src/main/java/com/facebook/react/bridge/PromiseImpl.java

```diff
@@ -5,67 +5,234 @@
+    reject(code, /*Message*/null, /*Throwable*/null, userInfo);
+  }
+
+  /**
+   * Report an exception with a custom code and userInfo.
+   *
+   * @param code      String
+   * @param throwable Throwable
+   * @param userInfo  WritableMap
+   */
+  @Override
+  public void reject(String code, Throwable throwable, WritableMap userInfo) {
+    reject(code, /*Message*/null, throwable, userInfo);
+  }
+
+  /**
+   * Report an error with a custom code, error message and userInfo,
+   * an error not caused by an exception.
+   *
+   * @param code     String
+   * @param message  String
+   * @param userInfo WritableMap
+   */
+  @Override
+  public void reject(String code, String message, @Nonnull WritableMap userInfo) {
+    reject(code, message, /*Throwable*/null, userInfo);
+  }
+
+  /**
+   * Report an exception with a custom code, error message and userInfo.
+   *
+   * @param code      String
+   * @param message   String
+   * @param throwable Throwable
+   * @param userInfo  WritableMap
+   */
+  @Override
+  public void reject(
+      @Nullable String code,
+      @Nullable String message,
+      @Nullable Throwable throwable,
+      @Nullable WritableMap userInfo
+  ) {
+    if (mReject == null) return;
+    WritableNativeMap errorInfo = new WritableNativeMap();
+
+    if (code == null) {
+      errorInfo.putString(ERROR_MAP_KEY_CODE, ERROR_DEFAULT_CODE);
+    } else {
+      errorInfo.putString(ERROR_MAP_KEY_CODE, code);
+    }
+
+    // Use the custom message if provided otherwise use the throwable message.
+    if (message != null) {
+      errorInfo.putString(ERROR_MAP_KEY_MESSAGE, message);
+    } else if (throwable != null) {
+      errorInfo.putString(ERROR_MAP_KEY_MESSAGE, throwable.getMessage());
+    } else {
+      // The JavaScript side expects a map with at least an error message.
+      // /Libraries/BatchedBridge/NativeModules.js -> createErrorFromErrorData
+      // TYPE: (errorData: { message: string })
+      errorInfo.putString(ERROR_MAP_KEY_MESSAGE, ERROR_DEFAULT_MESSAGE);
+    }
+
+    // For consistency with iOS ensure userInfo key exists, even if we null it.
+    // iOS: /React/Base/RCTUtils.m -> RCTJSErrorFromCodeMessageAndNSError
+    if (userInfo != null) {
+      errorInfo.putMap(ERROR_MAP_KEY_USER_INFO, userInfo);
+    } else {
+      errorInfo.putNull(ERROR_MAP_KEY_USER_INFO);
+    }
+
+    // Attach a nativeStackAndroid array if a throwable was passed
+    // this matches iOS behavior - iOS adds a `nativeStackIOS` property
+    // iOS: /React/Base/RCTUtils.m -> RCTJSErrorFromCodeMessageAndNSError
+    if (throwable != null) {
+      StackTraceElement[] stackTrace = throwable.getStackTrace();
+      WritableNativeArray nativeStackAndroid = new WritableNativeArray();
+
+      // Build an an Array of StackFrames to match JavaScript:
+      // iOS: /Libraries/Core/Devtools/parseErrorStack.js -> StackFrame
+      for (int i = 0; i < ERROR_STACK_FRAME_LIMIT; i++) {
+        StackTraceElement frame = stackTrace[i];
```

Comment:
Chance of IndexOutOfBoundsException if stacktrace is shorter than 10 elements?

Reply:
Hmm maybe, I'll update this to iterate to the lowest of the two: ERROR_STACK_FRAME_LIMIT or stackTrace.length

Thanks for the review 👌